### ISSUE TITLE
Member enforcement conflicts with GitLab-generated bots

### ISSUE THREAD

**Issue by lhokktyn** [gitlabform/gitlabform#454]
**Describe the bug**

> NOTE: I'm not sure if this is strictly a _bug_, but it does highlight a potential blocker in terms of using Project Access Tokens

Summary:

Setting `members.enforce: true` throws a 403 if there are project bot users associated with the project, i.e.

```shell
Request url='https://.../group/project/members/123456', method=DELETE, failed - expected code(s) [204, 404], got code 403 & body: '{"message":"403 Forbidden"}'
```

Replicate with:

1. Create a Project Access Token on the project to manage with GitLabForm (which in turn creates a bot user against that project)
2. Run `gitlabform` against that project with a config along the lines of ...

```yaml
config_version: 3

projects_and_groups:
  "path/to/project":
    members:
      users:
        example.real.user:
          access_level: maintainer
      enforce: true
```

**Desired behaviour**

I feel that GitLabForm should either ignore bot users when it comes to the `enforce` option, or allow an alternative option to let the user decide if bots should be ignored or not, e.g.

```yaml
config_version: 3

projects_and_groups:
  "path/to/project":
    members:
      users:
        example.real.user:
          access_level: maintainer
      enforce:
        named_members: true
        bot_members: true
```

**GitLabForm version**

3.3.2

**GitLab version**

`15.7.0-pre dc5804a9a33` (GitLab SaaS)

**Comment by gdubicki** [comment]
Hi @lhokktyn! Thanks for reporting this.

I agree that it would be more convenient to have a flag that would allow ignoring the bot users even when we want to enforce members.

Until this is implemented you can just add those users to the config - this is what we do at Egnyte. Example config:

```yaml
projects_and_groups:
  group_x/project_y:
    members:
      users:
        project_4187_bot:
          access_level: maintainer
      enforce: true
      # we actually manage the people how are members with `group_members:` section
      # that also contains `enforce: true`
```

We are open to reviewing and accepting PRs with this feature but we don't currently plan to implement it ourselves. One note about the syntax: I suggest a separate flag without changing how `enforce: true` works to make it backward compatible. Example:

```yaml
projects_and_groups:
  "path/to/project":
    members:
      users:
        example.real.user:
          access_level: maintainer
      enforce: true
      ignore_bot_members: true
```

**Comment by lhokktyn** [comment]
Thanks for the suggestion @gdubicki. I'll put a PR together.

**Comment by amimas** [comment]
> you can add those (bot) users to the config

That's what I'm doing too.

I'm not sure that there should be exceptions for enforcement. But if there is, I think it'd make sense for the config syntax to be a sub-key under `enforce`. But that means it's probably not going to be backward compatible. Depends on how it's implemented I guess.

I think what would be nice is if gitlabform could create the project access token (like deploy keys) and then we refer to that access token in the members section. What makes that complicated though is how to communicate that value of the access token. It probably shouldn't be accessible via stdout in gitlab ci or through a ci artifact. But that's a separate feature request :)

**Comment by adam-moss** [comment]
I'm curious how you manage churn in this scenario?  GitLab doesn't allow for the bot token to be regenerated, only revoked, so how are you managing expiry?

**Comment by amimas** [comment]
In my case I don't expect a lot of these bot users in various projects created via project access token.

I have a separate user (used for automation) created at an instance level. A personal access token with read only api scope is generated. This token is then added as CI variable at a group level. This way every project has access to a token that can be used for non-write type of api calls.

For projects that needs more tokens with more elevated permission, a token is created from the corresponding project. So it requires a bit of coordination. This does not happen often and since it's scoped to the project itself, haven't had a need for dealing with expiry...so far, because these tokens are typically used within gitlab ci of that same project.

Hope that helps. But you bring up a good point about dealing with these bots with expiry dates, if someone is setting expiration. I wonder how gitlabform deals with regular members with expiration when using the enforcement 🤔

**Comment by adam-moss** [comment]
Just FYI @amimas in case you have seen it in v16:

>In GitLab 16.0, any [personal](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html), [project](https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html), or [group](https://docs.gitlab.com/ee/user/group/settings/group_access_tokens.html) access token that does not have an expiration date will automatically have an expiration date set at one year.
>
>We recommend giving your access tokens an expiration date in line with your company's security policies before the default is applied:
>
>- On GitLab.com during the 16.0 milestone.
>- On GitLab self-managed instances when they are upgraded to 16.0.

**Comment by gdubicki** [comment]
Implemented in [v3.6.0](https://github.com/gitlabform/gitlabform/releases/tag/v3.6.0).
